# nf-core/metatdenovo: Output

## Introduction

This document describes the output produced by the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [Summary tables folder](#summary-tables-folder) - Final tables that can be imported directly in R
- [Preprocessing](#preprocessing)
  - [FastQC](#fastqc) - Read quality control
  - [Trimgalore](#trimgalore) - Primer trimming
  - [MultiQC](#multiqc) - Aggregate report describing results
  - [BBduk](#bbduk) - Filter out sequences from samples based on a fasta file (optional)
  - [BBnorm](#bbnorm) - Normalize the reads in the samples for a better assembly output (optional)
- [Assembly step](#assembly-step) - Generate contigs with an assembler program
  - [Megahit](#megahit) - Output from Megahit assembly (default)
  - [RNASpades](#rnaspades) - Output from Spades assembly (optional)
- [Orf Caller](#orf-caller) - Generate amino acids fasta file with an orf caller program
  - [Prodigal](#prodigal) - Output from Prodigal (default)
  - [Prokka](#prokka) - Output from Prokka (optional)
  - [TransDecoder](#transdecoder) - Output from transdecoder (optional) 
- [Functional and taxonomical annotation](#functional-and-taxonomical-annotation) - Predict the function out of the amino acids fasta file
  - [HMMER](#hmmer) - Analysis made with Hmmr profiles
  - [EggNOG](#eggnog) - Run EggNOG-mapper on amino acids fasta file
  - [KOfamSCAN](#kofamscan) - Run KOfamSCAN on amino acids fasta file
  - [EUKulele](#eukulele) - Run taxonomical annotation on amino acids fasta file
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### Summary tables folder

A summary report for all statistics results in tsv format. The report gives a general overview of the analysis, includes featureCounts output, taxonomical and functional annotation tables.
<details markdown="1">
<summary>Output file</summary>

- `summary_tables/`
  - `overall_stats.tsv`: statistics summary report.
  - `*counts.tsv`: summary table for featureCounts outputs
  - `*.tsv`: several tables based on different combination of the pipeline. From taxonomical to functional annotation (optional)

</details>

### Preprocessing

#### FastQC

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/). FastQC runs in trimgalore therefore its output can be found in Trimgalore's folder.

<details markdown="1">
<summary>Output files</summary>

- `trimgalore/fastqc/`
  - `*_fastqc.html`: FastQC report containing quality metrics for your untrimmed raw fastq files.

</details>

#### Trimgalore

[Trimgalore](https://github.com/FelixKrueger/TrimGalore) is trimming primer sequences from sequencing reads. Primer sequences are non-biological sequences that often introduce point mutations that do not reflect sample sequences. This is especially true for degenerated PCR primer. If primer trimming would be omitted, artifactual amplicon sequence variants might be computed by the denoising tool or sequences might be lost due to become labelled as PCR chimera.

<details markdown="1">
<summary>Output files</summary>

- `trimgalore/`: directory containing log files with retained reads, trimming percentage, etc. for each sample.
  - `*trimming_report.txt`: Report  of read numbers that pass trimgalore.

</details>

#### MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

:::note
The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.
:::

#### BBduk

[BBduk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbnorm-guide/) is a filtering tool that removes specific sequences from the samples using a reference fasta file.
BBduk is built-in tool from BBmap

<details markdown="1">
<summary>Output files</summary>

- `bbmap/`
  - `*.bbduk.log`: a text file with the results from BBduk analysis. Number of filtered reads can be seen in this log.

</details>

#### BBnorm

[BBnorm](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/) is a tool from BBmap that allows to reduce the coverage of highly abundant sequences and remove the sequences that are below a threshold, and can be useful if the data set is too large to assemble but also potentially improve an assembly. N.B. the digital normalization is done only for the assembly and the non-normalized sequences will be used for quantification
BBnorm is built-in tool from BBmap

<details markdown="1">
<summary>Output files</summary>

- `bbmap/bbnorm/logs/`
  - `*.logs`: it is a log file of the bbnorm run

</details>

### Assembly step

#### Megahit

[Megahit](https://github.com/voutcn/megahit) is used to assemble the cleaned and trimmed FastQ reads to create the reference genome.

<summary>Output file</summary>
- `megahit/megahit_out/`
  - `*.log`: it is a log file of Megahit run.
  - `megahit_assembly.contigs.fa.gz`: Reference genome created by Megahit.
  - `intermediate_contigs`: Folder that contains the intermediate steps of Megahit run.
    
</details>

#### RNASpades

Optionally, you can use RNASpades to assemble your reference genome. 
NB: we reccomend to use this assembler for eukaryotes rathern then prokaryotes.

<details markdown="1">
<summary>Output files</summary>

- `rnaspades/`
  - `rnaspades.assembly.gfa.gz`: gfa file output from rnaspades
  - `rnaspades.spades.log`: log file output from rnaspades run
  - `rnaspades.transcripts.fa.gz`: Reference genome created by RNASpades
</details>

### Orf caller

#### Prodigal

As default, you can use Prodigal to find ORFs on your reference genome. 

<details markdown="1">
<summary>Output files</summary>

- `prodigal/`
  - `*.fna.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

#### Prokka

As one alternative, you can use Prokka to find ORFs on your reference genome. 
NB: Prodigal and Prokka are reccomended for prokaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `prokka/`
  - `*.ffn.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

#### TransDecoder

Another alternative is TransDecoder to find ORFs on your reference genome. 
TransDecoder is reccomended for Eukaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `transdecoder/`
  - `*.cds`: nucleotides fasta file output
  - `*.pep`: amino acids fasta file output
  - `*.gff3`: genome feature file output

</details>

### Functional and taxonomical annotation

#### Hmmer
As one alternative, you can use Prokka to find ORFs on your reference genome. 
NB: Prodigal and Prokka are reccomended for prokaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `prokka/`
  - `*.ffn.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

#### EggNOG
As one alternative, you can use Prokka to find ORFs on your reference genome. 
NB: Prodigal and Prokka are reccomended for prokaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `prokka/`
  - `*.ffn.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

#### KOfamScan
As one alternative, you can use Prokka to find ORFs on your reference genome. 
NB: Prodigal and Prokka are reccomended for prokaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `prokka/`
  - `*.ffn.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

#### EUKulele
As one alternative, you can use Prokka to find ORFs on your reference genome. 
NB: Prodigal and Prokka are reccomended for prokaryotic samples

<details markdown="1">
<summary>Output files</summary>

- `prokka/`
  - `*.ffn.gz`: nucleotides fasta file output
  - `*.faa.gz`: amino acids fasta file output
  - `*.gff.gz`: genome feature file output

</details>

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>
